File: app/src/main/java/com/xiaomei/assistant/uihost/BaseSettingFragment.java
package com.xiaomei.assistant.uihost;

import android.content.Context;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import com.xiaomei.assistant.HostSettingsActivity;

public abstract class BaseSettingFragment extends Fragment {

    private HostSettingsActivity mSettingsHostActivity = null;
    @Nullable
    private String mTitle = null;
    @Nullable
    private String mSubtitle = null;

    @Override
    public void onAttach(@NonNull Context context) {
        super.onAttach(context);
        mSettingsHostActivity = (HostSettingsActivity) requireActivity();
    }

    @Override
    public void onDetach() {
        super.onDetach();
        mSettingsHostActivity = null;
    }

    @Nullable
    protected HostSettingsActivity getSettingsHostActivity() {
        return mSettingsHostActivity;
    }

    @NonNull
    protected HostSettingsActivity requireSettingsHostActivity() {
        if (mSettingsHostActivity == null) {
            throw new IllegalStateException("Host settings activity is null");
        }
        return mSettingsHostActivity;
    }

    public void finishFragment() {
        requireSettingsHostActivity().finishFragment(this);
    }

    @Nullable
    public String getTitle() {
        return mTitle;
    }

    @Nullable
    public String getSubtitle() {
        return mSubtitle;
    }

    protected void setTitle(@Nullable String title) {
        mTitle = title;
        if (mSettingsHostActivity != null) {
            mSettingsHostActivity.requestInvalidateActionBar();
        }
    }

    protected void setSubtitle(@Nullable String subtitle) {
        mSubtitle = subtitle;
        if (mSettingsHostActivity != null) {
            mSettingsHostActivity.requestInvalidateActionBar();
        }
    }
}

File: app/src/main/java/com/xiaomei/assistant/settings/BaseSettingFragment.kt
package com.xiaomei.assistant.settings

import android.content.Context
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment

abstract class BaseSettingFragment : Fragment() {
  private var settingsHostActivity: SettingsHostControllerOwner? = null
  private var currentTitle: String? = null
  private var currentSubtitle: String? = null

  override fun onAttach(context: Context) {
    super.onAttach(context)
    settingsHostActivity = SettingsHostOwnerRegistry.resolve(parentFragmentManager)
      ?: requireActivity() as? SettingsHostControllerOwner
      ?: error("Settings host owner is null: ${requireActivity().javaClass.name}")
  }

  override fun onDetach() {
    super.onDetach()
    settingsHostActivity = null
  }

  protected fun requireSettingsHostActivity(): SettingsHostControllerOwner {
    return settingsHostActivity ?: error("Settings host activity is null")
  }

  fun finishFragment() {
    requireSettingsHostActivity().finishFragment(this)
  }

  var titleText: String?
    get() = currentTitle
    protected set(value) {
      currentTitle = value
      settingsHostActivity?.requestInvalidateActionBar()
    }

  var subtitleText: String?
    get() = currentSubtitle
    protected set(value) {
      currentSubtitle = value
      settingsHostActivity?.requestInvalidateActionBar()
    }

  open fun notifyLayoutPaddingsChanged() {
  }

  open fun isWrapContent(): Boolean = true

  open fun doOnCreateView(
    inflater: LayoutInflater,
    container: ViewGroup?,
    savedInstanceState: Bundle?
  ): View? {
    return super.onCreateView(inflater, container, savedInstanceState)
  }

  final override fun onCreateView(
    inflater: LayoutInflater,
    container: ViewGroup?,
    savedInstanceState: Bundle?
  ): View? {
    return doOnCreateView(inflater, container, savedInstanceState)
  }

  open fun layoutPaddingTop(): Int = requireSettingsHostActivity().layoutPaddingTop
  open fun layoutPaddingBottom(): Int = requireSettingsHostActivity().layoutPaddingBottom
}

